pkg/image/registry_credentials_test.go

```diff
@@ -0,0 +1,118 @@
+package image
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+)
+
+func TestRegistryCredentials_Authenticator(t *testing.T) {
+	const exampleUsername = "some-example-username"
+	const examplePassword = "some-example-password"
+	const exampleToken = "some-example-token"
+
+	tests := []struct {
+		name                   string
+		credentials            RegistryCredentials
+		authenticatorAssertion func(t *testing.T, actual authn.Authenticator)
+	}{
+		{
+			name: "basic auth",
+			credentials: RegistryCredentials{
+				Username: exampleUsername,
+				Password: examplePassword,
+			},
+			authenticatorAssertion: basicAuth(authn.Basic{
+				Username: exampleUsername,
+				Password: examplePassword,
+			}),
+		},
+		{
+			name: "basic auth without username",
+			credentials: RegistryCredentials{
+				Username: "",
+				Password: examplePassword,
+			},
+			authenticatorAssertion: nilAuthenticator(),
+		},
+		{
+			name: "basic auth without password",
+			credentials: RegistryCredentials{
+				Username: exampleUsername,
+				Password: "",
+			},
+			authenticatorAssertion: nilAuthenticator(),
+		},
+		{
+			name: "bearer token",
+			credentials: RegistryCredentials{
+				Token: exampleToken,
+			},
+			authenticatorAssertion: bearerToken(authn.Bearer{
+				Token: exampleToken,
+			}),
+		},
+		{
+			name: "basic auth preferred over bearer token",
+			credentials: RegistryCredentials{
+				Username: exampleUsername,
+				Password: examplePassword,
+				Token:    exampleToken,
+			},
+			authenticatorAssertion: basicAuth(authn.Basic{
+				Username: exampleUsername,
+				Password: examplePassword,
+			}),
+		},
+		{
+			name:                   "no values provided",
+			credentials:            RegistryCredentials{},
+			authenticatorAssertion: nilAuthenticator(),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			actual := test.credentials.authenticator()
+			test.authenticatorAssertion(t, actual)
+		})
+	}
+}
+
+func basicAuth(expected authn.Basic) func(*testing.T, authn.Authenticator) {
+	return func(t *testing.T, actual authn.Authenticator) {
+		assertBasicAuth(t, expected, actual)
+	}
+}
+
+func bearerToken(expected authn.Bearer) func(*testing.T, authn.Authenticator) {
+	return func(t *testing.T, actual authn.Authenticator) {
+		assertBearerToken(t, expected, actual)
+	}
+}
+
+func nilAuthenticator() func(*testing.T, authn.Authenticator) {
+	return func(t *testing.T, actual authn.Authenticator) {
+		assert.Nil(t, actual)
+	}
+}
+
+func assertBasicAuth(t *testing.T, expected authn.Basic, actual authn.Authenticator) {
```

Comment:
why these don't use `t.Helper()` ?

Reply:
@alfredodeza Because I forgot! Awesome catch!

Calling `t.Helper()` would be nice here (and also in the corresponding assertion functions), because it makes the line reported to the user this: `test.authenticatorAssertion(t, actual)`, which is in the actual test code.

(The full error trace is still reported regardless of calls to `t.Helper()`, so it's easy to see where in the helper functions this bubbled up from.)

I shall update this!

Reply:
Updated here: https://github.com/anchore/stereoscope/pull/69/commits/3deea0cfd6231293c04d5028cf9324f32297b790